### PR TITLE
note added about Cycle and Permutation application of non-disjoint cycles

### LINKED ISSUE #12487: `Permutation` and `Cycle` as a callables have confusing behavior
There was some cleverness in the design of `Permutation` as a callable to mimic cycle notation.  However, I think this is really confusing.

Cycle notation:  when a permutation is written as a product of cycles, for example "(a,b)(c,d)" and you wish to apply it to a list `l`, that generally means

```
l[c], l[d] = l[d], l[c]
l[a], l[b] = l[b], l[a]
```

That is, first do the swap `(a,b)` then do the swap `(c,d)`.  The `Permutation` object cleverly uses callable to mimic this notation.  That is, you can construct an object `Permutation(a,b)(c,d)` since `Permutation(a,b)` is callable.  However, this seems contradictory to how functions are normally treated in a programming language.

When I encounter a function like `( Perm(swap a,b) )(swap c,d)` I expect that first `a,b` are swapped and then `c,d` are swapped which is equivalent in the current codebase to `Permutation(c,d)(a,b)`.  I think at minimum there needs to be a `kwarg` to allow the reversed order that I expect, but I also think that using callable this way is too clever a use of a programming grammar to look like a common mathematical notation.

### PR DESCRIPTION
**Pull request by smichr**
Old notes hidden in the comments of this OP.

No change has been made other than to put a note of caution in the docstring of permutation.

closes #12487

### CODE REVIEW

**Comment by smichr** [comment]
ping @siefkenj

**Comment by siefkenj** [comment]
I'm finding this pretty hard to follow...The functionality I need is the following:  A call `Permuation(some_list).list() == some_list` provided that `some_list` is a permutation and a call `Permutation(list_of_tuples).list()` applies the cycles from left-to-right.  Maybe I am misreading the code, but it seems like `Cycle(some_list)` applies `some_list` as a cycle, not as an explicitly specified permutation as `Permutation` would.

**Comment by smichr** [comment]
Here's a demo:

```
>>> from sympy.combinatorics import *
>>> Permutation([1,2,3,5,0,4]).list()
[1, 2, 3, 5, 0, 4]
>>> l2r = Permutation([1,2], [2,3])  # multi-lists is NEW
>>> r2l = Permutation([[1,2], [2,3]])  # list of lists (old R to L evaluation)
>>> l2r == Permutation(2,3)(1,2)
True
>>> r2l == Permutation(1,2)(2,3)
True
```

**Comment by siefkenj** [comment]
Okay.  I still don't think this can be used for my purposes, though.  If someone gives me a list of numbers, it is an explicit permutation, and if someone gives me a list of pairs, it is a list of swaps when evaluated left to right.  So, if someone gives me `w`, I cannot do `Permutation(w)`, nor can I do `Permutation(*w)`, which isn't to comment on whether this is a worthwhile addition to the `Permutation` class, but it doesn't free me from going a bunch of manual argument checking.

**Comment by smichr** [comment]
> but it doesn't free me from going a bunch of manual argument checking.

See point (2) in the OP. With that you could do `Permutation(w, from_left=True)` if the `from_left` were only used to handle the list of lists case. But I would prefer to just fix this via that pathway given in the second sentence of point 2. What do you think about that?

In the meanwhile, I think `Permutation(w if not w or not is_sequence(w[0]) else list(reversed(w)))` should work.

**Comment by jksuom** [comment]
It is often confusing that there are two different ways of composing permutations. This is how the product of permutations is defined in SymPy:
> Return the product a*b as a Permutation; the ith value is b(a(i)).

However, permutations are also mappings and, when considered as mappings, the composed mapping whose value at i is b(a(i)) is denoted b°a (with degree sign emulating the proper \circ sign).

Hence 'left to right' using one composition means 'right to left' with the other composition. So the ordering of cycles in this example
```
        >>> C((2, 3), (1, 2))
        (1 3 2)
        >>> C((1, 2), (2, 3))
        (1 2 3)
```
is opposite to that of composed permutations
```
>>> p = Permutation(Cycle(1, 2))
>>> q = Permutation(Cycle(2, 3))
>>> q*p  # ordered as in C((2, 3), (1, 2))
Permutation(1, 2, 3)
>>> p*q
Permutation(1, 3, 2)
```

**Comment by jksuom** [comment]
I'm afraid that having two methods of composing permutations would lead to confusion and coding errors. The definition currently used in SymPy is prevalent in the literature, and it is the one that those working in this area are expecting. That is also the order in which permutations are returned by some matrix methods. The alternative order does not seem to be used in SymPy. Hence I would prefer keeping a single order as in calls like `Cycle(a, b)(c, d)` (where a, b, c and d need not be all different).

**Comment by smichr** [comment]
> The definition currently used in SymPy is prevalent in the literature

Glad for that reassurance.

OK, then, perhaps just adding a docstring example of the `*` multiplication and the `Cycle` notation would suffice. 

```
>>> a = Cycle(1, 2)
>>> b = Cycle(2, 3)
>>> Permutation(a)*Permutation(b) == Cycle(1,2)(2,3)
True
```

In [this comment] I already demonstrate how the result for @siefkenj 's case can be handled.

**Comment by asmeurer** [comment]
> I'm afraid that having two methods of composing permutations would lead to confusion and coding errors. The definition currently used in SymPy is prevalent in the literature, and it is the one that those working in this area are expecting. That is also the order in which permutations are returned by some matrix methods. The alternative order does not seem to be used in SymPy. Hence I would prefer keeping a single order as in calls like Cycle(a, b)(c, d) (where a, b, c and d need not be all different).

I completely agree with this.

**Comment by smichr** [comment]
I will simply add a note in the docstring about the ordering issue and that it only matters when the cycles are not disjoint. Or maybe the docs would be a better place for this?

Here is a workaround for @siefkenj :+1: 
```
>>> def LRPerm(w):
...     from sympy.utilities.iterables import is_sequence
...     return Permutation(w if not (w and is_sequence(w[0])) else
...     Permutation.rmul(*[Permutation(*i) for i in w]))
>>> LRPerm(((1,2),(2,3)))  # cycle input
(123)
>>> Permutation(2,3)(1,2)
(123)
>>> LRPerm((1,2,0))  # actual-permutation input
(012)
```

**Comment by smichr** [comment]
If someone can see if this added note is clear, then this is done.
